beacon_node/lighthouse_network/src/rpc/mod.rs

```diff
@@ -348,57 +376,60 @@ where
-                            } else {
-                                // Other protocols shouldn't be sending large messages, we should flag the peer kind
-                                crit!(self.log, "Request size too large to ever be processed"; "protocol" => %protocol);
-                            }
-                            // send an error code to the peer.
-                            // the handler upon receiving the error code will send it back to the behaviour
-                            self.send_response(
-                                peer_id,
-                                (conn_id, *id),
-                                RPCCodedResponse::Error(
-                                    RPCResponseErrorCode::RateLimited,
-                                    "Rate limited. Request too large".into(),
-                                ),
-                            );
-                        }
-                        Err(RateLimitedErr::TooSoon(wait_time)) => {
-                            debug!(self.log, "Request exceeds the rate limit";
-                        "request" => %req, "peer_id" => %peer_id, "wait_time_ms" => wait_time.as_millis());
-                            // send an error code to the peer.
-                            // the handler upon receiving the error code will send it back to the behaviour
-                            self.send_response(
-                                peer_id,
-                                (conn_id, *id),
-                                RPCCodedResponse::Error(
-                                    RPCResponseErrorCode::RateLimited,
-                                    format!("Wait {:?}", wait_time).into(),
-                                ),
-                            );
+                if !self.active_inbound_requests_limiter.allows(
+                    peer_id,
+                    req.versioned_protocol().protocol(),
+                    &conn_id,
+                    id,
+                ) {
+                    // There is already an active request with the same protocol. Send an error code to the peer.
+                    debug!(self.log, "There is an active request with the same protocol"; "peer_id" => peer_id.to_string(), "request" => %req, "protocol" => %req.versioned_protocol().protocol());
+                    self.send_response(
+                        peer_id,
+                        (conn_id, *id),
+                        RPCCodedResponse::Error(
+                            RPCResponseErrorCode::RateLimited,
+                            "Rate limited. There is an active request with the same protocol"
+                                .into(),
+                        ),
+                    );
+                    return;
+                }
+
+                if let Some(limiter) = self.inbound_request_size_limiter.as_ref() {
+                    // Check if the request is conformant to the quota
+                    if limiter.allows(req) {
```

Comment:
Shouldn't the limiter here be peer specific here? Do we want rate limiting here if multiple peers collectively cross the limit?
For e.g. for a ping request which is 2 requests every 10 secs, if we get > 2 peers sending us ping requests in a 10 sec window, the third peer will get rate limited here even though it is its first ping request

Reply:
`self.inbound_request_size_limiter` (`RequestSizeLimiter`)  looks similar to the `Limiter`, but it just checks if the request is too large; it doesn't perform rate-limiting. So, this doesn't need to be peer-specific.


For reference, here is the reason why I introduced `RequestSizeLimiter`, quoted from [the PR comment](https://github.com/sigp/lighthouse/pull/5923#issue-2351213755):

> This request size check is also performed in `Before` using the `Limiter`, but in this PR, I introduced `RequestSizeLimiter` to handle this. Unlike the `Limiter`, `RequestSizeLimiter` is dedicated to perform this check.
>
> The reasons why I introduced `RequestSizeLimiter` are:
>
> - In `After`, the rate limiter is shared between the behaviour and the handler (`Arc<Mutex<RateLimiter>>>`). (This is detailed at the next sentence)
> - The request size check does not heavily depend on the rate-limiting logic, so it can be separated with minimal code duplication.
> - The request size check is performed on the behaviour side. By separating the request size check from the rate limiter, we can reduce the locking of the rate limiter.

Reply:
Ohh thanks for explaining. 

In that case, can't we vastly simplify the entire check by just doing
```rust
match request.protocol() {
            Protocol::Ping
            | Protocol::Status
            | Protocol::Goodbye
            | Protocol::MetaData
            | Protocol::LightClientBootstrap
            | Protocol::LightClientOptimisticUpdate
            | Protocol::LightClientFinalityUpdate
            | Protocol::BlocksByRoot
            | Protocol::BlobsByRoot => true,
            Protocol::BlocksByRange => request.max_responses() > spec.max_request_blocks(),
            Protocol::BlobsByRange => request.max_responses() > spec.max_blobs(),
        }
```

For the roots requests, the RuntimeVariable ssz list ensures that we don't get more requests than the max specified in the config. This way we don't need to bother with the tokening system for checking request sizes

Reply:
That makes sense!

I’m working on the simplification you suggested, but I couldn’t find `spec.max_blobs()`. Could you provide any advice on this? 🙏

Reply:
Sorry i meant `max_request_blob_sidecars`

Reply:
https://github.com/sigp/lighthouse/pull/5923/commits/5a9237fdc6a5ad11281ec7c36b2b7bdfdeba7ed4

I removed the `RequestSizeLimiter` and added `is_request_size_too_large()` instead, which does the check. That looks much simpler. ✨ Thanks for your suggestion.